File: codebase/apps/backend/src/db/my-data.ts
import { count, desc, eq } from "drizzle-orm";
import { monthToDateUsd } from "../ai/meter";
import { schema, type Db } from "./client";

// "My data" (FR-3.15, design §7 /me/data): what the system stores about one user, read-only
// and owner-scoped. Never the password hash, never a token.

const RECENT = 50;

export async function myData(db: Db, userId: string) {
  const user = await db.query.users.findFirst({ where: eq(schema.users.id, userId) });
  if (!user) return null;

  const profileVersions = await db
    .select({ version: schema.profiles.version, status: schema.profiles.status, createdAt: schema.profiles.createdAt })
    .from(schema.profiles)
    .where(eq(schema.profiles.userId, userId))
    .orderBy(desc(schema.profiles.version));
  const gateChoices = await db
    .select({ gate: schema.gateChoices.gate, choice: schema.gateChoices.choice, freeText: schema.gateChoices.freeText, source: schema.gateChoices.source, chosenAt: schema.gateChoices.chosenAt })
    .from(schema.gateChoices)
    .where(eq(schema.gateChoices.userId, userId))
    .orderBy(desc(schema.gateChoices.chosenAt))
    .limit(RECENT);
  const editDiffs = await db
    .select({ draftId: schema.editDiffs.draftId, diff: schema.editDiffs.diff, createdAt: schema.editDiffs.createdAt })
    .from(schema.editDiffs)
    .where(eq(schema.editDiffs.userId, userId))
    .orderBy(desc(schema.editDiffs.createdAt))
    .limit(RECENT);
  const revisions = await db
    .select({ draftId: schema.draftRevisions.draftId, revisionNo: schema.draftRevisions.revisionNo, instructions: schema.draftRevisions.instructions, createdAt: schema.draftRevisions.createdAt })
    .from(schema.draftRevisions)
    .innerJoin(schema.drafts, eq(schema.drafts.id, schema.draftRevisions.draftId))
    .where(eq(schema.drafts.userId, userId))
    .orderBy(desc(schema.draftRevisions.createdAt))
    .limit(RECENT);
  const draftsByStatus = await db
    .select({ status: schema.drafts.status, n: count() })
    .from(schema.drafts)
    .where(eq(schema.drafts.userId, userId))
    .groupBy(schema.drafts.status);
  const limits = await db.query.userLimits.findFirst({ where: eq(schema.userLimits.userId, userId) });

  return {
    account: {
      email: user.email,
      displayName: user.displayName,
      role: user.role,
      sanityProjectId: user.sanityProjectId,
      sanityDataset: user.sanityDataset,
      createdAt: user.createdAt,
      lastActiveAt: user.lastActiveAt,
    },
    profileVersions,
    gateChoices,
    editDiffs,
    revisions,
    draftsByStatus,
    spend: {
      monthToDateUsd: Number((await monthToDateUsd(db, userId)).toFixed(4)),
      monthlyCapUsd: Number(limits?.monthlyCapUsd ?? 10),
      maxRunsPerDay: limits?.maxRunsPerDay ?? 2,
      autoPublish: limits?.autoPublish ?? false,
    },
  };
}

File: codebase/apps/mobile/lib/screens/my_data.dart
import 'package:flutter/material.dart';

import '../api.dart';

/// "My data" (FR-3.15): what the system stores about you, read-only. Deletion is an
/// administrator action (FR-2.6).
class MyDataScreen extends StatefulWidget {
  const MyDataScreen({super.key});

  @override
  State<MyDataScreen> createState() => _MyDataScreenState();
}

class _MyDataScreenState extends State<MyDataScreen> {
  Map<String, dynamic>? _data;
  String? _error;

  @override
  void initState() {
    super.initState();
    ApiClient.instance.get('/me/data').then(
          (d) => setState(() => _data = d),
          onError: (Object e) => setState(() => _error = '$e'),
        );
  }

  String _date(Object? raw) => raw == null ? '—' : DateTime.parse(raw as String).toLocal().toString().substring(0, 16);

  List<Widget> _section(String title, List<dynamic> rows, String Function(Map<String, dynamic>) line) => [
        Padding(
          padding: const EdgeInsets.only(top: 20, bottom: 6),
          child: Text('$title (${rows.length})', style: Theme.of(context).textTheme.titleMedium),
        ),
        if (rows.isEmpty) const Text('Nothing stored.'),
        for (final r in rows.cast<Map<String, dynamic>>())
          Padding(padding: const EdgeInsets.only(bottom: 4), child: Text(line(r), maxLines: 3, overflow: TextOverflow.ellipsis)),
      ];

  @override
  Widget build(BuildContext context) {
    final d = _data;
    final account = d?['account'] as Map<String, dynamic>?;
    final spend = d?['spend'] as Map<String, dynamic>?;
    return Scaffold(
      appBar: AppBar(title: const Text('My data')),
      body: _error != null
          ? Center(child: Text(_error!))
          : d == null
              ? const Center(child: CircularProgressIndicator())
              : ListView(padding: const EdgeInsets.all(16), children: [
                  const Text('Read-only. To have your data deleted, ask an administrator.'),
                  ..._section('Account', [account!], (a) =>
                      '${a['displayName']} · ${a['email']} · ${a['role']}\nSite: ${a['sanityProjectId'] ?? '—'} · since ${_date(a['createdAt'])}'),
                  ..._section('Spend this month', [spend!], (s) =>
                      '\$${s['monthToDateUsd']} of \$${s['monthlyCapUsd']} · ${s['maxRunsPerDay']} runs/day · auto-publish ${s['autoPublish'] == true ? 'on' : 'off'}'),
                  ..._section('Profile versions', d['profileVersions'] as List<dynamic>,
                      (p) => 'v${p['version']} · ${p['status']} · ${_date(p['createdAt'])}'),
                  ..._section('Connected accounts', (d['socialAccounts'] as List<dynamic>?) ?? [],
                      (a) => '${a['provider']} · ${a['handle']} · ${a['state']} · connected ${_date(a['connectedAt'])}'),
                  ..._section('Drafts', d['draftsByStatus'] as List<dynamic>, (r) => '${r['status']}: ${r['n']}'),
                  ..._section('Choices you made', d['gateChoices'] as List<dynamic>,
                      (g) => '${g['gate']} · ${g['source']}${g['freeText'] != null ? ' · "${g['freeText']}"' : ''} · ${_date(g['chosenAt'])}'),
                  ..._section('Revision instructions', d['revisions'] as List<dynamic>,
                      (r) => '#${r['revisionNo']} "${r['instructions']}" · ${_date(r['createdAt'])}'),
                  ..._section('Your edits (diffs)', d['editDiffs'] as List<dynamic>,
                      (e) => '${_date(e['createdAt'])}\n${e['diff']}'),
                ]),
    );
  }
}
